File: src/test/diversion_syscall.c
/* -*- Mode: C; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */

#include "util.h"

static __attribute__((noinline)) void breakpoint(void) {
  int break_here = 1;
  (void)break_here;
}

int main(void) {
  // We will make use of the fact that fd 3 is closed. Just to be sure the
  // parent didn't leak anything, close it (in the normal case, this'll fail,
  // but that's ok).
  close(3);

  // Do a bunch of bufferable system calls
  int fd = open("/dev/null", O_WRONLY | O_CLOEXEC);
  test_assert(fd == 3);

  // At this breakpoint, we'll attempt to open a file during the diversion
  breakpoint();

  // More bufferable syscalls
  for (int i = 1; i < 10; ++i) {
    write(fd, "Hello", 5);
  }
}

File: src/test/diversion_syscall.py
from util import *
import re

send_gdb('b breakpoint')
expect_gdb('Breakpoint 1')

send_gdb('c')
expect_gdb('Breakpoint 1')

send_gdb('p (int)open("diversion_print.out",0102,0600)')
expect_gdb(r'\$1 = (\d+)')
fd = last_match().group(1)
# This FD is confused with that during record. That's ok, just make sure
# operations on this during the diversion actually go to our file.

send_gdb(f'p (int)write({fd}, "DIVERSION-SUCCESS", 17)')
expect_gdb(r'\$2 = 17')

ok()
